### PR TITLE
Workaround some gcc 13.2.0+ warnings.

### LINKED ISSUE #15631: Compiler error from migrating the StableHLO input conversion to a compiler plugin.
### What happened?

When pulling after commit commit ebb5b7d4a3c59b8dac9c5a4fe4ba78f0c736dc7f
I get the following build error.
/usr/local/google/home/lubol/iree/compiler/plugins/input/StableHLO/stablehlo-iree/Conversion/StableHLOToLinalgPointwise.cpp:261:63:   required from here
/usr/local/google/home/lubol/iree/third_party/llvm-project/mlir/include/mlir/IR/PatternMatch.h:254:16: error: ‘virtual void mlir::RewritePattern::rewrite(mlir::Operation*, mlir::PatternRewriter&) const’ was hidden [-Werror=overloaded-virtual=]
  254 |   virtual void rewrite(Operation *op, PatternRewriter &rewriter) const;
      |                ^~~~~~~
/usr/local/google/home/lubol/iree/third_party/llvm-project/mlir/include/mlir/Transforms/DialectConversion.h:544:16: note:   by ‘mlir::OpConversionPattern<mlir::stablehlo::XorOp>::rewrite’
  544 |   virtual void rewrite(SourceOp op, OpAdaptor adaptor,
      |                ^~~~~~~
cc1plus: all warnings being treated as errors
[5913/7172] Building CXX object compil...ib/Dialect/Torch/IR/TorchDialect.cpp.o
ninja: build stopped: subcommand failed.

### Steps to reproduce your issue

Clone the latest IREE repo and do a default build. The result is the error above.

### What component(s) does this issue relate to?

_No response_

### Version information

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by ScottTodd**
Fixes https://github.com/openxla/iree/issues/15631 . We recommend using clang / MSVC instead of gcc (for _many_ reasons), but many Linux distributions default to using gcc, so it helps to have a clean/successful build there. Our gcc build bot on CI is running gcc 9, while 13+ is available.

## `-Wno-overloaded-virtual`

At least gcc 13.2.0 reports more warnings with this than clang, like these:

```
/usr/local/google/home/lubol/iree/compiler/plugins/input/StableHLO/stablehlo-iree/Conversion/StableHLOToLinalgPointwise.cpp:261:63: required from here
/usr/local/google/home/lubol/iree/third_party/llvm-project/mlir/include/mlir/IR/PatternMatch.h:254:16: error: ‘virtual void mlir::RewritePattern::rewrite(mlir::Operation*, mlir::PatternRewriter&) const’ was hidden [-Werror=overloaded-virtual=]
254 | virtual void rewrite(Operation *op, PatternRewriter &rewriter) const;
| ^~~~~~~
/usr/local/google/home/lubol/iree/third_party/llvm-project/mlir/include/mlir/Transforms/DialectConversion.h:544:16: note: by ‘mlir::OpConversionPatternmlir::stablehlo::XorOp::rewrite’
544 | virtual void rewrite(SourceOp op, OpAdaptor adaptor,
| ^~~~~~~
cc1plus: all warnings being treated as errors
[5913/7172] Building CXX object compil...ib/Dialect/Torch/IR/TorchDialect.cpp.o
```

The warnings may be legit, but I can't really figure out how to resolve them, and other compilers (with the same warning) don't complain.

## `<1>` in types

These produce warnings like this:

```
[build] /usr/include/c++/13/bits/stl_algobase.h:437:30: error: ‘void* __builtin_memmove(void*, const void*, long unsigned int)’ forming offset 24 is out of the bounds [0, 24] of object ‘<anonymous>’ with type ‘llvm::SmallVector<long int, 1>’ [-Werror=array-bounds=]
[build]   437 |             __builtin_memmove(__result, __first, sizeof(_Tp) * _Num);
[build]       |             ~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

### CODE REVIEW

**Comment by ScottTodd** [comment]
Well, I tried to test this locally, but my machine ran out of disk space and I'm about to go on vacation 🙃

File: build_tools/cmake/iree_copts.cmake [hunk 1]
```diff
@@ -212,6 +212,7 @@ iree_select_compiler_opts(IREE_DEFAULT_COPTS
     "-Wno-comment"
     "-Wno-format-zero-length"
     "-Wno-uninitialized"
+    "-Wno-overloaded-virtual"  # gcc complains more than clang, see #15631
```
Comment (ScottTodd) [inline build_tools/cmake/iree_copts.cmake]: Argh, found another warning/error:
```
[build] /usr/local/google/home/scotttodd/code/iree-build/_deps/spirv-tools-src/source/opt/trim_capabilities_pass.cpp:146:17: error: ‘*(unsigned int*)((char*)&<unnamed> + offsetof(std::optional<spv::Capability>,std::optional<spv::Capability>::<unnamed>.std::_Optional_base<spv::Capability, true, true>::<unnamed>))’ may be used uninitialized [-Werror=maybe-uninitialized]
[build]   146 |     return std::nullopt;
[build]       |                 ^~~~~~~
```

I wonder if the compiler plugin changes affected which flags we apply to third party code. Probably need to look at `iree_cc_library`, `external_cc_library` (which has no uses right now?), etc.?

**Review by LLITCHEV — APPROVED** [review]
LGTM

File: build_tools/cmake/iree_copts.cmake (hunk 1 above)
Comment (LLITCHEV) [reply build_tools/cmake/iree_copts.cmake]: I got the following error when testing locally...
cc1: error: command-line option ‘-Wno-overloaded-virtual=0’ is valid for C++/ObjC++ but not for C [-Werror]

Comment (LLITCHEV) [reply build_tools/cmake/iree_copts.cmake]: Yeah, it sounds to me that building the 3rd-party libs with the iree build warning flags is going to create warning/errors during build. They  are having different warning level settings. I think...

Comment (ScottTodd) [reply build_tools/cmake/iree_copts.cmake]: The other warnings are sort of tracked here: https://github.com/openxla/iree/issues/11394

I had enabled `-DIREE_TARGET_BACKEND_WEBGPU=ON` (default disabled), which does not build with gcc warnings as errors right now.

Comment (ScottTodd) [reply build_tools/cmake/iree_copts.cmake]: > I got the following error when testing locally... cc1: error: command-line option ‘-Wno-overloaded-virtual=0’ is valid for C++/ObjC++ but not for C [-Werror]

Whoops, thanks. I changed this to only add the flag for C++.

**Review by LLITCHEV — APPROVED** [review]
LGTM

**Pull request merged**